### ISSUE TITLE
Caching issue: Templates seem to disappear till restart of DNN

### ISSUE THREAD

**Issue by iJungleboy** [2sic/2sxc#1564]
**I'm submitting a ...**  
[x] bug report => search github for a similar issue before submitting 

**...about**   
[x] edit experience / UI

**Current behavior**
We're not sure yet, but it seems sometimes templates and content-blocks become invisible in the app until DNN is restarted. I presume it's a caching issue. 

**Minimal reproduction of the problem with instructions**
Cannot reliably reproduce it yet. 

**Please tell us about your environment:**

* **2sxc version(s):** 9.30

**Comment by iJungleboy** [comment]
This is hard to reproduce. Last time we had it we saw that the admin-ui of the app could open, but didn't show any views/templates or content-blocks (in the hidden `2SexyContent-System` scope. Restarting DNN returned everything to normal. 

ATM I have a few guesses:

1. either the lookup table app-guid to app-id gets broken - so it actually looks at the wrong app, resulting in not finding the template (which is not on the app it's trying to access)
1. the cache gets changed and has a mistake in partial-rebuilding, resulting in some data missing
1. there is a full cache-rebuild, but for some reasons these are not loaded (maybe because they are global-types)

ATM I'm guessing it's 1, but we'll have to see it again when it breaks. I'll try to add more logging or find some kind of method to debug it next time we see this.

**Comment by iJungleboy** [comment]
Since we can't reliably reproduce it yet, I probably won't have a fix in 9.31, but I'm creating a deep-insights debug view to look into the data-model as is in memory to better track it down #1565. I hope we'll have that on an installation when we next see something like this...

**Comment by iJungleboy** [comment]
fyi: i'm starting to find the issue, thanks to a new feature I'm building which gives insight to the internal caches. I believe it's related to the global types - i'll keep you posted...

**Comment by iJungleboy** [comment]
I believe I fixed it - the cause would have been this:

1. an initializer which only runs once loads global content-types
1. if it fails, it doesn't try again
1. it used Server.MapPath to find paths
1. If the search-index of DNN hit it first, it would fail because Server didn't exist (search runs in a background thread)

So if the first access to an app happens through a search-indexer, it will try to load some things - fail - and never try again. I changed it to use HostingEnvironment.MapPath and believe it works.
